src/proto/h1/conn.rs

```diff
@@ -217,24 +226,10 @@ where
         trace!("Conn::read_head");
 
         #[cfg(feature = "server")]
-        if !self.state.h1_header_read_timeout_running {
-            if let Some(h1_header_read_timeout) = self.state.h1_header_read_timeout {
-                let deadline = Instant::now() + h1_header_read_timeout;
-                self.state.h1_header_read_timeout_running = true;
-                match self.state.h1_header_read_timeout_fut {
-                    Some(ref mut h1_header_read_timeout_fut) => {
-                        trace!("resetting h1 header read timeout timer");
-                        self.state.timer.reset(h1_header_read_timeout_fut, deadline);
-                    }
-                    None => {
-                        trace!("setting h1 header read timeout timer");
-                        self.state.h1_header_read_timeout_fut =
-                            Some(self.state.timer.sleep_until(deadline));
-                    }
-                }
-            }
-        }
```

Comment:
I was forced to move the code that starts the timeout after the parsing is polled, because which timeout to use depends on whether the connection is idle (no bytes read on all poll attempts) or not (some bytes read on at least one poll attempt).

My earlier attempts to predict whether bytes would be available based on the read buffer failed (the read buffer looks empty until the first poll of the io in parsing).

I also considered starting the idle timeout elsewhere, but the code would be more fragile as a single rouge poll of read_head would start the header read timer when it shouldn't be started.